R/get-path-from-object.R

```diff
@@ -158,45 +158,234 @@ get_model_id.bbi_model <- function(.mod) {
+}
+
+#' @rdname get_data_path
+#' @export
+get_data_path.bbi_log_df <- function(
+    .bbi_object,
+    .check_exists = TRUE,
+    ...
+){
+  data_paths <- map_chr(.bbi_object[[ABS_MOD_PATH]], function(.path) {
+    .mod <- read_model(.path)
+    get_data_path(.mod, .check_exists, ...)
+  })
+
+  return(data_paths)
+}
+
+
+get_data_path_nonmem <- function(
+    .bbi_object,
+    .check_exists = TRUE,
+    ...
+){
+  # Get data path as defined in control stream
+  data_path_ctl <- get_data_path_from_ctl(.bbi_object)
+  # Get data path as defined in bbi_config.json
+  data_path_config <- tryCatch(
+    get_data_path_from_json(.bbi_object),
+    error = function(.e) {
+      if(stringr::str_detect(.e$message, "Cannot extract data path")){
+        return(NULL)
+      }else{stop(.e)}
+    }
+  )
+
+  if(is.null(data_path_config)) {
+    data_path <- data_path_ctl
+  }else{
+    # if they're not the same, warn
+    if(data_path_config != data_path_ctl){
+      mod_path <- get_model_path(.bbi_object)
+      rlang::warn(
+        c(
+          paste("Data path referenced in `bbi_config.json` does not match the",
+                "one defined in the control stream:"),
+          "i" = glue("Data path extracted from {basename(mod_path)}: {data_path_ctl}"),
+          "i" = glue("Data path extracted from bbi_config.json: {data_path_config}"),
+          "Preferring config (`bbi_config.json`) path..."
+        )
+      )
+    }
+    data_path <- data_path_config
+  }
+
+
+  if(.check_exists){
+    if(!fs::file_exists(data_path)){
+      # The first error message line is what NMTRAN would return in this situation
+      rlang::abort(
+        c(
+          "x" = "Input data file does not exist or cannot be opened",
+          "i" = glue("Referenced input data path: {data_path}")
+        )
+      )
+    }
+  }
+
+  return(data_path)
+}
+
+
+
+#' Get the data path from a control stream file
+#' @noRd
+get_data_path_from_ctl <- function(.mod, normalize = TRUE){
+  check_model_object(.mod, "bbi_nonmem_model")
+  mod_path <- get_model_path(.mod)
+  ctl <- nmrec::read_ctl(mod_path)
+
+  # Get data record
+  data_recs <- nmrec::select_records(ctl, "data")
+  n_data <- length(data_recs)
+  if(n_data !=1){
+    recs_fmt <- purrr::map_chr(data_recs, function(rec) rec$format())
+    rlang::abort(
+      c(
+        glue::glue("Expected a single data record, but found {n_data}:\n\n"),
+        glue("{recs_fmt}")
+      )
+    )
+  }
+
+  # Get file path and remove quotes
+  data_path_ctl <- nmrec::get_record_option(data_recs[[1]], "filename")$value
+  data_path_ctl <- unquote_filename(data_path_ctl)
+
+  if(isTRUE(normalize)){
+    # Normalize to current model directory
+    model_dir <- dirname(get_model_path(.mod, .check_exists = FALSE))
+    # Adjust for model extension
+    data_path_ctl_adj <- adjust_data_path_ext(data_path_ctl, mod_path)
+    data_path <- fs::path_norm(file.path(model_dir, data_path_ctl_adj)) %>%
+      as.character()
```

Comment:
Won't this mishandle absolute paths?

For example, for an absolute path "/baz/zzz" this logic would incorrectly append the model directory:

```R
fs::path_norm(file.path("/foo/bar", "/baz/zzz"))
#> /foo/bar/baz/zzz
```

(`get_data_path_from_json` may have a similar pre-existing bug here; it depends on whether bbi remaps a data path to a relative path before writing `bbi_config.json` and I haven't confirmed that yet.)

Reply:
> it depends on whether bbi remaps a data path to a relative path before writing `bbi_config.json` and I haven't confirmed that yet

I just checked and it looks like bbi mishandles it and errors:

```
mdev:bbi-cli-testing$ grep DATA abs.ctl
$DATA /data/home/kylem/scratch/bbi-cli-testing/acop.csv IGNORE=@
mdev:bbi-cli-testing$ bbi nonmem run local --overwrite abs.ctl
INFO[0000] Successfully loaded default configuration from /data/home/kylem/scratch/bbi-cli-testing/bbi.yaml 
INFO[0000] Beginning Local Path                         
INFO[0000] A total of 1 models have completed the initial preparation phase 
INFO[0000] [abs] Beginning local work phase             
INFO[0007] [abs] Beginning cleanup phase                
ERRO[0007] abs error reading data to hash: open /data/home/kylem/scratch/bbi-cli-testing/abs/data/home/kylem/scratch/bbi-cli-testing/acop.csv: no such file or directory 
ERRO[0007] abs error hashing data: invalid argument     
INFO[0007] [abs] Cleanup completed                      
INFO[0007] Waiting for any post execution hooks to finish 
1 models completed in 7.64932725s       
```

Which of course makes a fix here less important, but I'd still appreciate if this was adjusted to not append the model's directory to an absolute path.

Reply:
It at least ends up with the correct absolute path in the `bbi_config.json`, but there's no hash:

```
$  jq '{data_path, data_md5}' <abs/bbi_config.json 
{
  "data_path": "/data/home/kylem/scratch/bbi-cli-testing/acop.csv",
  "data_md5": ""
}
```

Summary: it would be great if both the `from_ctl` and `from_json` avoided tacking on an invalid leading path to an absolute file, but there are already existing issues here, so I don't consider that a blocker.

Given the above issue,

Reply:
How do you envision knowing if the path is absolute or not? I tried something like the following, but it caused a number of tests to fail:
```r
    # Adjust for model extension
    data_path_ctl_adj <- adjust_data_path_ext(data_path_ctl, mod_path)
    # Normalize to current model directory if not an absolute path
    model_dir <- dirname(get_model_path(.mod, .check_exists = FALSE))
    if(data_path_ctl_adj != normalizePath(data_path_ctl_adj)){
      data_path <- fs::path_norm(file.path(model_dir, data_path_ctl_adj)) %>%
        as.character()
    }else{
      data_path <- data_path_ctl_adj
    }
```

Reply:
> How do you envision knowing if the path is absolute or not?

`fs::is_absolute_path()`?   If that's true, avoid doing `file.path(model_dir, ...)`.

Reply:
That seemed to work more appropriately. Added a test for this as well

Reply:
I said:

> I just checked and it looks like bbi mishandles it and errors when hashing the file

Issue filed here: https://github.com/metrumresearchgroup/bbi/issues/316